src/tools/pub/cmd/pub-add.md

```diff
@@ -26,65 +26,149 @@ For example, if 0.13.3 is the latest stable version of the `http` package,
 then `dart pub add http` adds
 `http: ^0.13.3` under `dependencies` in the pubspec.
 
-To add a [dev dependency][], use the `--dev` option:
-
-[dev dependency]: /tools/pub/dependencies#dev-dependencies
+You can also specify a constraint or constraint range
 
 ```terminal
-$ dart pub add --dev test
+$ dart pub add foo:2.0.0
+$ dart pub add foo:'^2.0.0'
+$ dart pub add foo:'>2.0.0 <3.0.1'
 ```
 
-## Options
+If `dart pub add <package>:<constraint>` is an existing dependency,
+it will update the constraint.
 
-For options that apply to all pub commands, see
-[Global options](/tools/pub/cmd#global-options).
+## YAML descriptor
 
-{{site.alert.note}}
-  Any specified options will apply to all the packages
-  included in an invocation of the command.
-  For example, `dart pub add test http --dev` 
-  will add both the `test` and `http` packages 
-  as dev dependencies.
+{{site.alert.version-note}}
+  YAML-formatted descriptor syntax was added in Dart 2.19.
+  The descriptor replaces arguments like `--path`, `--sdk`, `--git-<option>`, etc.
+  We still support these arguments, but the documented method is now YAML-descriptor only.
 {{site.alert.end}}
 
-### `-d, --dev`
+The YAML descriptor reflects how dependencies are written in `pubspec.yaml`.
+
+```nocode
+$ dart pub add [options] [dev:]<package>[:descriptor] [[dev:]<package>[:descriptor]
+       ...]
+```
+
+Within the `descriptor` you can add a package with specific constraints or other sources:
+```nocode
+'<package>:{"<source>":"<descriptor>"[,"<source>":"<descriptor>"],"version":"<constraint>"}'
+```
+
+The `descriptor` syntax cannot be used in conjunction with any of the optional arguments it replaces.
+Their new corresponding YAML sources are listed below.
 
-Adds the package as a dev dependency,
+### `dev`
```

Comment:
It seems `dev` is not really part of the YAML descriptor, but rather a separate piece, similar to the version constraint above. Do you think it should be above this section?

Reply:
Oh right, I had the same thought and had originally named the YAML section just "YAML _syntax_" that would include `dev`, and then a subsection called "descriptors" under that with all the options. But yeah for the sake of not having a million section headers on a relatively short page, I'll move `dev` out. Good catch!

Reply:
On second thought, I'm going to leave it where it is and instead rename the section to "YAML syntax" like I originally intended because:
- I want to keep the first section new-syntax agnostic for the time being, to clearly define what's new and what's not
- I don't want to introduce `dev:` outside of the context of the rest of the new syntax, and then introduce the new syntax, that'd be weird
- The line at the end of the descriptor section intro: 
  > The `descriptor` syntax cannot be used in conjunction with any of the optional
arguments it replaces.

  is true for `dev:` too, except `dev:` is not part of the descriptor. So I'll just change "descriptor" to something more general in that sentence.

Reply:
Hmm, hard. I do see value in the new functionality being collocated, but one concern is `dev` is not part of the YAML syntax and it's located in a different location. YAML/JSON is just the format of source descriptor after the colon. Do you think its conclusion would cause confusion or would it be fine for now?

I also think we should try to mention "descriptor" in the section header somehow, since the section doesn't describe "YAML syntax" as a whole or in general, but rather a specific usage of it. Perhaps "Source descriptor" or "Source descriptor syntax" then describe it uses a YAML syntax?

Reply:
Oh I see. I thought `dev:` wasn't part of the yaml _descriptor_, but it's not yaml at all. Ok let's go with your first idea then, moving dev up the page before the yaml section. Thanks for the extra explanation!